api/v1alpha1/healthcheck_types.go

```diff
@@ -0,0 +1,149 @@
+// Copyright Envoy Gateway Authors
+// SPDX-License-Identifier: Apache-2.0
+// The full text of the Apache license is available in the LICENSE file at
+// the root of the repo.
+
+package v1alpha1
+
+import metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+// HealthCheck defines the health check configuration.
+type HealthCheck struct {
+	// Timeout defines the time to wait for a health check response.
+	//
+	// +kubebuilder:validation:Format=duration
+	Timeout metav1.Duration `json:"timeout"`
+
+	// Interval defines the time between health checks.
+	//
+	// +kubebuilder:validation:Format=duration
+	Interval metav1.Duration `json:"interval"`
+
+	// UnhealthyThreshold defines the number of unhealthy health checks required before a backend host is marked unhealthy.
+	//
+	// +kubebuilder:validation:Minimum=1
+	UnhealthyThreshold uint32 `json:"unhealthyThreshold"`
+
+	// HealthyThreshold defines the number of healthy health checks required before a backend host is marked healthy.
+	//
+	// +kubebuilder:validation:Minimum=1
+	HealthyThreshold uint32 `json:"healthyThreshold"`
+
+	// HealthChecker defines the concrete health checker to do health checking.
+	HealthChecker HealthChecker `json:"healthChecker"`
+}
+
+// HealthCheckerType is the type of health checker.
+// +kubebuilder:validation:Enum=HTTP;GRPC;TCP
+type HealthCheckerType string
+
+const (
+	// HealthCheckerTypeHTTP defines the HTTP type of health checking.
+	HealthCheckerTypeHTTP HealthCheckerType = "HTTP"
+	// HealthCheckerTypeGRPC defines the GRPC type of health checking.
+	HealthCheckerTypeGRPC HealthCheckerType = "GRPC"
+	// HealthCheckerTypeTCP defines the TCP type of health checking.
+	HealthCheckerTypeTCP HealthCheckerType = "TCP"
+)
+
+// HealthChecker defines the configuration of concrete health checker.
+// EG supports various types of health checking including HTTP, GRPC, TCP.
+// +union
+//
+// +kubebuilder:validation:XValidation:rule="self.type == 'HTTP' ? has(self.http) : !has(self.http)",message="If Health Checker type is HTTP, http field needs to be set."
+// +kubebuilder:validation:XValidation:rule="self.type == 'GRPC' ? has(self.grpc) : !has(self.grpc)",message="If Health Checker type is GRPC, grpc field needs to be set."
+// +kubebuilder:validation:XValidation:rule="self.type == 'TCP' ? has(self.tcp) : !has(self.tcp)",message="If Health Checker type is TCP, tcp field needs to be set."
+type HealthChecker struct {
+	// Type defines the type of health checker.
+	// +kubebuilder:validation:Enum=HTTP;GRPC;TCP
+	// +unionDiscriminator
+	Type HealthCheckerType `json:"type" yaml:"type"`
```

Comment:
is there a lot of value in splitting up HTTP and GRPC ?
imo 2 types should be sufficient where response is checked for HTTP and conn establishment is checked for TCP

Reply:
prefer if we only implement HTTP and TCP for now

Reply:
OK